api/aggregation.proto

```diff
@@ -0,0 +1,181 @@
+    // matches, the match configuration matches.
+    MatchSet and_match = 1;
+
+    // A set that describes a logical OR. If all members of the set
+    // match, the match configuration matches.
+    MatchSet or_match = 2;
+
+    // A negation match. The match configuration will match if the
+    // negated match condition matches.
+    MatchPredicate not_match = 3;
+
+    // The match configuration will always match.
+    bool any_match = 4 [(validate.rules).bool = {const: true}];
+
+    // Match on an Envoy request type.
+    // Ex: "type.googleapis.com/envoy.api.v2.Listener"
+    RequestTypeMatch request_type_match = 5;
+
+    // Match on a Field in Envoy's request node.
+    RequestNodeMatch request_node_match = 6;
+  }
+}
+
+// Rules for how to generate the resulting fragment of the xDS Aggregator cache
+// key.
+// [#next-free-field: 5]
+message ResultPredicate {
+
+  message ResultAction {
+
+    // TODO potentially use "safe regex"
+    // https://github.com/envoyproxy/envoy/blob/10f756efa17e56c8d4d1033be7b4286410db4e01/api/envoy/type/matcher/v3/regex.proto
+    // [#next-free-field: 3]
+    message RegexAction {
+
+      string pattern = 1 [(validate.rules).message.required = true];
+
+      string replace = 2 [(validate.rules).message.required = true];
+    }
+
+    oneof action {
+      option (validate.required) = true;
+
+      // Uses the exact string value of the field.
+      bool exact = 1 [(validate.rules).bool = {const: true}];
+
+      // Operates a regex find and replace on the field.
+      RegexAction regex_action = 2;
+    }
+  }
+
+  // Rules for generating the resulting fragment from a Envoy request node.
+  // [#next-free-field: 3]
+  message RequestNodeFragment {
+
+    NodeFieldType field = 1 [(validate.rules).message.required = true];
+
+    ResultAction action = 2 [(validate.rules).message.required = true];
+  }
+
+  // Rules for generating the resulting fragment from Envoy request names.
+  // [#next-free-field: 4]
+  message ResourceNamesFragment {
+
+    NodeFieldType field = 1 [(validate.rules).message.required = true];
+
+    // The index of the resource name to operate on.
+    // TODO support operation on multiple elements once we move past the MVP.
+    int32 element = 2 [(validate.rules).message.required = true];
+
+    ResultAction action = 3 [(validate.rules).message.required = true];
+  }
+
+  oneof type {
+    option (validate.required) = true;
+
+    // A set that describes a logical AND. The result is a non-separated
+    // append operation between two or more fragments.
+    RepeatedResultPredicate result_predicate = 1 ;
+
+    // A fragment generated from an Envoy request node.
+    RequestNodeFragment request_node_fragment = 2;
+
+    // A fragment generated from Envoy resource names.
+    ResourceNamesFragment resource_names_fragment = 3;
+
+    // A simple string fragment
+    string string_fragment = 4;
+  }
+
+  message RepeatedResultPredicate {
```

Comment:
This message is the only real difference between the spec and what the proto compiler accepts. I was seeing this error in:
```
aggregation.proto:164:9: Fields in oneofs must not have labels (required / optional / repeated).
```